### PR TITLE
inherit [index.lifecycle.date] from rolled-over time

### PR DESCRIPTION
**Pull request by talevy**
Indices that are rolled over preserve their origin index.lifecycle.date
as the time the index was created. Although this is also fine, it would
make more sense to start the timer for moving to the warm phase from the
time it was rolled over so that we capture the time elapsed from the
latest data, as opposed to when the index was created.

This commit adds an extra step within the RolloverAction that
extracts the index.creation_date of the newly created index from the RolloverStep
and sets that as the index.lifecycle.date of the rolled-over index that is
waiting for its next phase.

### CODE REVIEW

**Comment by talevy** [comment]
I've opened https://github.com/elastic/elasticsearch/issues/30887 to discuss

**Comment by talevy** [comment]
Hi @colings86, I've updated this PR to leverage RolloverInfo

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UpdateRolloverLifecycleDateStep.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.rollover.RolloverInfo;
+import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+
+public class UpdateRolloverLifecycleDateStep extends AsyncActionStep {
```
Comment (colings86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UpdateRolloverLifecycleDateStep.java]: I'm not sure if we should have this as an async action step or have it work directly on the cluster state update thread like `PolicyInitialisationStep` does? It seems strange to me that we set this setting on the cluster state update thread initially and then later as an async action step.

Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UpdateRolloverLifecycleDateStep.java]: good catch, it was a reflex for updating settings... but this setting is controlled by us and safe to do on the cluster-state 👍

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/InitializePolicyContextStep.java [hunk 2]
```diff
@@ -11,13 +11,14 @@
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.Index;
 
-public class InitializePolicyContextStep extends Step {
+public class InitializePolicyContextStep extends ClusterStateActionStep {
```
Comment (colings86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/InitializePolicyContextStep.java]: As this step is special and we rely on it definitely running as is I think wee should make it final. 

Actually I think we should make all the steps final but that can be in a different change. I think its most important that this one is final as we need to ensure the creation date is definitely set properly

**Review by colings86 — APPROVED** [review]
Left a minor comment bbut LGTM

**Pull request merged**